File: app/src/main/java/com/clockmods/sdk/clock/ClockThemeTokens.java
package com.clockmods.sdk.clock;

import android.graphics.Typeface;

/**
 * Portable visual tokens shared by previews and renderers.
 *
 * <p>Structure and geometry intentionally remain in {@link ClockRenderer}; changing these tokens
 * alone is not considered a new clock style.</p>
 */
public final class ClockThemeTokens {
    private final int backgroundStartColor;
    private final int backgroundEndColor;
    private final int surfaceColor;
    private final int primaryTextColor;
    private final int secondaryTextColor;
    private final int accentColor;
    private final int lineColor;
    private final String displayFontFamily;
    private final String supportingFontFamily;
    private final Typeface displayTypeface;
    private final Typeface supportingTypeface;
    private final float strokeScale;

    private ClockThemeTokens(Builder builder) {
        backgroundStartColor = builder.backgroundStartColor;
        backgroundEndColor = builder.backgroundEndColor;
        surfaceColor = builder.surfaceColor;
        primaryTextColor = builder.primaryTextColor;
        secondaryTextColor = builder.secondaryTextColor;
        accentColor = builder.accentColor;
        lineColor = builder.lineColor;
        displayFontFamily = builder.displayFontFamily;
        supportingFontFamily = builder.supportingFontFamily;
        displayTypeface = builder.displayTypeface;
        supportingTypeface = builder.supportingTypeface;
        strokeScale = builder.strokeScale;
    }

    public static Builder builder() {
        return new Builder();
    }

    /** A builder pre-loaded with these tokens, for deriving a variant of an existing theme. */
    public Builder toBuilder() {
        return new Builder()
                .background(backgroundStartColor, backgroundEndColor)
                .surfaceColor(surfaceColor)
                .primaryTextColor(primaryTextColor)
                .secondaryTextColor(secondaryTextColor)
                .accentColor(accentColor)
                .lineColor(lineColor)
                .fonts(displayFontFamily, supportingFontFamily)
                .typefaces(displayTypeface, supportingTypeface)
                .strokeScale(strokeScale);
    }

    public int getBackgroundStartColor() { return backgroundStartColor; }
    public int getBackgroundEndColor() { return backgroundEndColor; }
    public int getSurfaceColor() { return surfaceColor; }
    public int getPrimaryTextColor() { return primaryTextColor; }
    public int getSecondaryTextColor() { return secondaryTextColor; }
    public int getAccentColor() { return accentColor; }
    public int getLineColor() { return lineColor; }
    public String getDisplayFontFamily() { return displayFontFamily; }
    public String getSupportingFontFamily() { return supportingFontFamily; }
    /**
     * A resolved typeface overriding {@link #getDisplayFontFamily()}, or {@code null} when the
     * theme's own family stands. The host sets this from the user's per-theme font and weight;
     * a style's own tokens never carry one, so previews and tests keep the theme's design intact.
     */
    public Typeface getDisplayTypeface() { return displayTypeface; }
    /** As {@link #getDisplayTypeface()}, for the supporting family. */
    public Typeface getSupportingTypeface() { return supportingTypeface; }
    public float getStrokeScale() { return strokeScale; }

    public static final class Builder {
        private int backgroundStartColor = 0xFF101418;
        private int backgroundEndColor = 0xFF080A0D;
        private int surfaceColor = 0xFF20262D;
        private int primaryTextColor = 0xFFFFFFFF;
        private int secondaryTextColor = 0xFFADB5BD;
        private int accentColor = 0xFFFF5A52;
        private int lineColor = 0xFF6B747E;
        private String displayFontFamily = "sans-serif";
        private String supportingFontFamily = "sans-serif";
        private Typeface displayTypeface;
        private Typeface supportingTypeface;
        private float strokeScale = 1f;

        public Builder background(int startColor, int endColor) {
            backgroundStartColor = startColor;
            backgroundEndColor = endColor;
            return this;
        }

        public Builder surfaceColor(int color) { surfaceColor = color; return this; }
        public Builder primaryTextColor(int color) { primaryTextColor = color; return this; }
        public Builder secondaryTextColor(int color) { secondaryTextColor = color; return this; }
        public Builder accentColor(int color) { accentColor = color; return this; }
        public Builder lineColor(int color) { lineColor = color; return this; }

        public Builder fonts(String displayFamily, String supportingFamily) {
            displayFontFamily = requireFamily(displayFamily);
            supportingFontFamily = requireFamily(supportingFamily);
            return this;
        }

        /**
         * Overrides both families with already-resolved typefaces. Either may be {@code null},
         * which leaves that slot on the theme's own family.
         */
        public Builder typefaces(Typeface display, Typeface supporting) {
            displayTypeface = display;
            supportingTypeface = supporting;
            return this;
        }

        public Builder strokeScale(float scale) {
            if (Float.isNaN(scale) || Float.isInfinite(scale) || scale <= 0f) {
                throw new IllegalArgumentException("strokeScale must be finite and positive");
            }
            strokeScale = scale;
            return this;
        }

        public ClockThemeTokens build() {
            return new ClockThemeTokens(this);
        }

        private static String requireFamily(String family) {
            if (family == null || family.trim().length() == 0) {
                throw new IllegalArgumentException("font family must not be blank");
            }
            return family.trim();
        }
    }
}

File: app/src/main/java/com/clockmods/sdk/clock/ClockThemeTokens.kt
package com.clockmods.sdk.clock

import android.graphics.Typeface

/** Portable visual tokens shared by previews and renderers. */
class ClockThemeTokens private constructor(
    backgroundStartColor: Int,
    backgroundEndColor: Int,
    surfaceColor: Int,
    primaryTextColor: Int,
    secondaryTextColor: Int,
    accentColor: Int,
    lineColor: Int,
    displayFontFamily: String,
    supportingFontFamily: String,
    displayTypeface: Typeface?,
    supportingTypeface: Typeface?,
    strokeScale: Float,
    gaussianBlur: Boolean,
    blurStrength: Int,
    blurBrightness: Int,
    cardShadow: Boolean,
) {
    companion object {
        const val DEFAULT_BLUR_STRENGTH = 50
        const val DEFAULT_BLUR_BRIGHTNESS = 25

        @JvmStatic
        fun builder() = Builder()
    }

    private val backgroundStartColor = backgroundStartColor
    private val backgroundEndColor = backgroundEndColor
    private val surfaceColor = surfaceColor
    private val primaryTextColor = primaryTextColor
    private val secondaryTextColor = secondaryTextColor
    private val accentColor = accentColor
    private val lineColor = lineColor
    private val displayFontFamily = displayFontFamily
    private val supportingFontFamily = supportingFontFamily
    private val displayTypeface = displayTypeface
    private val supportingTypeface = supportingTypeface
    private val strokeScale = strokeScale
    private val gaussianBlur = gaussianBlur
    private val blurStrength = blurStrength
    private val blurBrightness = blurBrightness
    private val cardShadow = cardShadow

    fun toBuilder() = Builder()
        .background(backgroundStartColor, backgroundEndColor)
        .surfaceColor(surfaceColor)
        .primaryTextColor(primaryTextColor)
        .secondaryTextColor(secondaryTextColor)
        .accentColor(accentColor)
        .lineColor(lineColor)
        .fonts(displayFontFamily, supportingFontFamily)
        .typefaces(displayTypeface, supportingTypeface)
        .strokeScale(strokeScale)
        .gaussianBlur(gaussianBlur)
        .blurStrength(blurStrength)
        .blurBrightness(blurBrightness)
        .cardShadow(cardShadow)

    fun getBackgroundStartColor() = backgroundStartColor
    fun getBackgroundEndColor() = backgroundEndColor
    fun getSurfaceColor() = surfaceColor
    fun getPrimaryTextColor() = primaryTextColor
    fun getSecondaryTextColor() = secondaryTextColor
    fun getAccentColor() = accentColor
    fun getLineColor() = lineColor
    fun getDisplayFontFamily() = displayFontFamily
    fun getSupportingFontFamily() = supportingFontFamily
    fun getDisplayTypeface() = displayTypeface
    fun getSupportingTypeface() = supportingTypeface
    fun getStrokeScale() = strokeScale
    fun isGaussianBlur() = gaussianBlur
    fun getBlurStrength() = blurStrength
    fun getBlurBrightness() = blurBrightness
    fun isCardShadow() = cardShadow

    class Builder {
        private var backgroundStartColor = 0xFF101418.toInt()
        private var backgroundEndColor = 0xFF080A0D.toInt()
        private var surfaceColor = 0xFF20262D.toInt()
        private var primaryTextColor = 0xFFFFFFFF.toInt()
        private var secondaryTextColor = 0xFFADB5BD.toInt()
        private var accentColor = 0xFFFF5A52.toInt()
        private var lineColor = 0xFF6B747E.toInt()
        private var displayFontFamily = "sans-serif"
        private var supportingFontFamily = "sans-serif"
        private var displayTypeface: Typeface? = null
        private var supportingTypeface: Typeface? = null
        private var strokeScale = 1f
        private var gaussianBlur = false
        private var blurStrength = DEFAULT_BLUR_STRENGTH
        private var blurBrightness = DEFAULT_BLUR_BRIGHTNESS
        private var cardShadow = false

        fun gaussianBlur(enabled: Boolean) = apply { gaussianBlur = enabled }
        fun cardShadow(enabled: Boolean) = apply { cardShadow = enabled }
        fun blurStrength(percent: Int) = apply { blurStrength = percent.coerceIn(0, 100) }
        fun blurBrightness(percent: Int) = apply { blurBrightness = percent.coerceIn(0, 100) }
        fun background(startColor: Int, endColor: Int) = apply {
            backgroundStartColor = startColor
            backgroundEndColor = endColor
        }
        fun surfaceColor(color: Int) = apply { surfaceColor = color }
        fun primaryTextColor(color: Int) = apply { primaryTextColor = color }
        fun secondaryTextColor(color: Int) = apply { secondaryTextColor = color }
        fun accentColor(color: Int) = apply { accentColor = color }
        fun lineColor(color: Int) = apply { lineColor = color }
        fun fonts(displayFamily: String?, supportingFamily: String?) = apply {
            displayFontFamily = requireFamily(displayFamily)
            supportingFontFamily = requireFamily(supportingFamily)
        }
        fun typefaces(display: Typeface?, supporting: Typeface?) = apply {
            displayTypeface = display
            supportingTypeface = supporting
        }
        fun strokeScale(scale: Float) = apply {
            require(scale.isFinite() && scale > 0f) { "strokeScale must be finite and positive" }
            strokeScale = scale
        }
        fun build() = ClockThemeTokens(
            backgroundStartColor,
            backgroundEndColor,
            surfaceColor,
            primaryTextColor,
            secondaryTextColor,
            accentColor,
            lineColor,
            displayFontFamily,
            supportingFontFamily,
            displayTypeface,
            supportingTypeface,
            strokeScale,
            gaussianBlur,
            blurStrength,
            blurBrightness,
            cardShadow,
        )

        private fun requireFamily(family: String?) = family?.trim().orEmpty().also {
            require(it.isNotEmpty()) { "font family must not be blank" }
        }
    }
}
